numalogic/udfs/inference.py

```diff
@@ -0,0 +1,225 @@
+        """
+        Register config with the UDF.
+
+        Args:
+            config_id: Config ID
+            conf: StreamConf object
+        """
+        self.pl_conf.stream_confs[config_id] = conf
+
+    def get_conf(self, config_id: str) -> StreamConf:
+        try:
+            return self.pl_conf.stream_confs[config_id]
+        except KeyError as err:
+            raise ConfigNotFoundError(f"Config with ID {config_id} not found!") from err
+
+    def compute(self, model: artifact_t, input_: npt.NDArray[float], **_) -> npt.NDArray[float]:
+        """
+        Perform inference on the input data.
+
+        Args:
+            model: Model artifact
+            input_: Input data
+
+        Returns
+        -------
+            Reconstruction error
+
+        Raises
+        ------
+            RuntimeError: If model forward pass fails
+        """
+        x = torch.from_numpy(input_).unsqueeze(0)
+        model.eval()
+        try:
+            with torch.no_grad():
+                _, out = model.forward(x)
+            recon_err = model.criterion(out, x, reduction="none")
+        except Exception as err:
+            raise RuntimeError("Model forward pass failed!") from err
+        return np.ascontiguousarray(recon_err).squeeze(0)
+
+    def exec(self, keys: list[str], datum: Datum) -> Messages:
+        """
+        Perform inference on the input data.
+
+        Args:
+            keys: List of keys
+            datum: Datum object
+
+        Returns
+        -------
+            Messages instance
+        """
+        _start_time = time.perf_counter()
+
+        # Construct payload object
+        payload = StreamPayload(**orjson.loads(datum.value))
+
+        _LOGGER.debug(
+            "%s - Received Msg: { CompositeKeys: %s, Metrics: %s }",
+            payload.uuid,
+            payload.composite_keys,
+            payload.metrics,
+        )
+
+        # Forward payload if a training request is tagged
+        if payload.header == Header.TRAIN_REQUEST:
+            return Messages(Message(keys=keys, value=payload.to_json()))
+
+        artifact_data = self.load_artifact(keys, payload)
+        # Send training request if artifact loading is not successful
+        if not artifact_data:
+            payload = replace(
+                payload, status=Status.ARTIFACT_NOT_FOUND, header=Header.TRAIN_REQUEST
+            )
+            return Messages(Message(keys=keys, value=payload.to_json()))
+
+        # Perform inference
+        try:
+            x_inferred = self.compute(artifact_data.artifact, payload.get_data())
+        except RuntimeError:
+            _LOGGER.exception(
+                "%s - Runtime inference error! Keys: %s, Metric: %s",
+                payload.uuid,
+                payload.composite_keys,
+                payload.metrics,
+            )
+            payload = replace(payload, status=Status.RUNTIME_ERROR, header=Header.TRAIN_REQUEST)
```

Comment:
No but the model inference failed. This could be due to a change in the config, e.g. we changed seq len from 10 to 16. In that case, it will throw a runtime error since the old model is not valid anymore with the new config, and we need a new model to be trained.

Reply:
Again, we need to revisit the situations where we want this to happen, and where we don't.